File: src/test/java/accounting/patterns/CustomerTest.java
package accounting.patterns;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Currency;

import static accounting.patterns.EntryType.BASE_USAGE;
import static accounting.patterns.EventType.USAGE;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;


class CustomerTest {

    @Test
    @DisplayName("Posting an event to a customer creates a valid entry")
    void testPostingAnEvent() {
        LocalDate effectiveDate = LocalDate.of(1999, 10, 1);
        Customer customer = new Customer("Acme Coffee Makers",
                new ServiceAgreement(new BigDecimal(10))
                        .addPostingRule(USAGE,
                                new MultiplyByRatePostingRule(BASE_USAGE),
                                effectiveDate));

        customer.post(USAGE, effectiveDate, new Quantity(50));

        Entry resultingEntry = customer.getEntry(0);
        assertEquals(new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(500)), resultingEntry.amount());
    }

    @Test
    @DisplayName("Posting an event to a customer without posting rule")
    void testCustomerWithNoRules() {
        LocalDate effectiveDate = LocalDate.of(1999, 10, 1);
        Customer customer = new Customer("Acme Coffee Makers",
                new ServiceAgreement(new BigDecimal(10)));

        assertThrows(RuntimeException.class, () -> customer.post(USAGE, effectiveDate, new Quantity(50)));
    }

    @Test
    @DisplayName("Customers have names")
    void customersAreNamed() {
        assertEquals("foo", new Customer("foo", new ServiceAgreement(new BigDecimal(1))).getName());
    }
}

File: src/test/typescript/accounting-pattern/CustomerTest.ts
import { Customer } from "../../main/typescript/accounting-pattern/Customer";
import { ServiceAgreement } from "../../main/typescript/accounting-pattern/ServiceAgreement";
import { EventType } from "../../main/typescript/accounting-pattern/EventType";
import { EntryType } from "../../main/typescript/accounting-pattern/EntryType";
import { MultiplyByRatePostingRule } from "../../main/typescript/accounting-pattern/MultiplyByRatePostingRule";
import { Quantity } from "../../main/typescript/accounting-pattern/Quantity";
import { MonetaryAmount } from "../../main/typescript/accounting-pattern/MonetaryAmount";
import { Entry } from "../../main/typescript/accounting-pattern/Entry";

describe("Customer", () => {
    it("should create a valid entry when posting an event", () => {
        const effectiveDate = new Date(1999, 9, 1);
        const customer = new Customer("Acme Coffee Makers",
            new ServiceAgreement(10)
                .addPostingRule(EventType.USAGE, new MultiplyByRatePostingRule(EntryType.BASE_USAGE), effectiveDate)
        );

        customer.post(EventType.USAGE, effectiveDate, new Quantity(50));

        const resultingEntry = customer.getEntry(0);
        expect(resultingEntry.getAmount()).toEqual(new MonetaryAmount("USD", 500));
    });

    it("should throw an error when posting an event without a posting rule", () => {
        const effectiveDate = new Date(1999, 9, 1);
        const customer = new Customer("Acme Coffee Makers", new ServiceAgreement(10));

        expect(() => customer.post(EventType.USAGE, effectiveDate, new Quantity(50))).toThrow();
    });

    it("should have a name", () => {
        expect(new Customer("foo", new ServiceAgreement(1)).getName()).toEqual("foo");
    });
});
